telegram/ext/_application.py

```diff
@@ -0,0 +1,1224 @@
+            if isinstance(exception, ApplicationHandlerStop):
+                warn(
+                    'ApplicationHandlerStop is not supported with asynchronously '
+                    'running handlers.',
+                    stacklevel=1,
+                )
+
+            # Avoid infinite recursion of error handlers.
+            elif is_error_handler:
+                _logger.exception(
+                    'An error was raised and an uncaught error was raised while '
+                    'handling the error with an error_handler.',
+                    exc_info=exception,
+                )
+
+            else:
+                # If we arrive here, an exception happened in the task and was neither
+                # ApplicationHandlerStop nor raised by an error handler.
+                # So we can and must handle it
+                await self.dispatch_error(update, exception, coroutine=coroutine)
+
+            # Raise exception so that it can be set on the task
+            raise exception
+        finally:
+            self._mark_for_persistence_update(update=update)
+
+    async def _update_fetcher(self) -> None:
+        # Continuously fetch updates from the queue. Exit only once the signal object is found.
+        while True:
+            update = await self.update_queue.get()
+
+            if update is _STOP_SIGNAL:
+                _logger.debug('Dropping pending updates')
+                while not self.update_queue.empty():
+                    self.update_queue.task_done()
+
+                # For the _STOP_SIGNAL
+                self.update_queue.task_done()
+                return
+
+            _logger.debug('Processing update %s', update)
+
+            if self._concurrent_updates:
+                self.create_task(self.__process_update_wrapper(update), update=update)
+            else:
+                await self.__process_update_wrapper(update)
+
+    async def __process_update_wrapper(self, update: object) -> None:
+        async with self._concurrent_updates_sem:
+            await self.process_update(update)
+            self.update_queue.task_done()
+
+    async def process_update(self, update: object) -> None:
+        """Processes a single update and updates the persistence.
+
+        .. versionchanged:: 14.0
+            This calls :meth:`update_persistence` exactly once after handling of the update was
+            finished by *all* handlers that handled the update, including asynchronously running
+            handlers.
+
+        Args:
+            update (:class:`telegram.Update` | :obj:`object` | \
+                :class:`telegram.error.TelegramError`):
+                The update to process.
+
+        """
+        context = None
+        any_blocking = False
+
+        for handlers in self.handlers.values():
+            try:
+                for handler in handlers:
+                    check = handler.check_update(update)
+                    if not (check is None or check is False):
```

Comment:
because `check` can be an abitrary object, it might evaluate to `False` in a boolean context without being `None` or `False` and the documented behavior or `check_update` is that only those two make the handler not handle the update (IMHO that's also more well-defined than "return anything that evaluates to `False`")